### PR TITLE
implement raw version of bluetooth connect feature

### PR DESCRIPTION
**Pull request by belens**
## What this PR changes
- Adds functionality to connect via bluetooth
- Added implementation to run local server with HTTPS (needed in my case so chrome allowed me to request bluetooth devices)
- 

## Rationale for the change
The LD2410 has strange data error when connected via USB. Therefor i was forced to make a bluetooth implementation for my project. This repo helped me make way better sense of the hardware docs (thanks!), so i thought i would implement this feature in the repository.

## Checklist

- [x ] The PR does one thing OR multiple closely-related things
  - If it does multiple unrelated things, please consider separate PRs
- [ x] Commits are in the present tense (e.g. `Display firmware version in header`)

### CODE REVIEW

**Comment by belens** [comment]
This PR is not ready for merging yet, I have some TODO's placed inline which i need to fix, as well as making the interfactor more usable. 

The reason I open the PR already it to verify that you are interested in this feature before I spend more time on it.

File: .vscode/settings.json [hunk 1]
```diff
@@ -6,5 +6,15 @@
...
+    "typescript",
+    "javascript",
+    "svelte"
+  ],
+  "[svelte]": {
+    "editor.defaultFormatter": "svelte.svelte-vscode"
+  },
+  "[typescript]": {
+    "editor.defaultFormatter": "vscode.typescript-language-features"
+  }
```
Comment (albertnis) [inline .vscode/settings.json:19]: Anything in particular which inspired this change? I'd rather use prettier, which is set up with the svelte plugin to work on `.svelte` files.

That said, if you're having issues with this setup, I'd love to move discussion and changes to a separate issue or PR and we can proceed from there.

File: vite.config.ts [hunk 2]
```diff
@@ -1,6 +1,12 @@
 import { sveltekit } from "@sveltejs/kit/vite";
 import { defineConfig } from "vite";
+import mkcert from 'vite-plugin-mkcert';
 
 export default defineConfig({
-  plugins: [sveltekit()],
+  plugins: [sveltekit(), mkcert()],
```
Comment (albertnis) [inline vite.config.ts:6]: I confess I haven't tested this, but my understanding was that [`localhost` is a secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure) so would therefore function fine with APIs such as Bluetooth. Web Serial also requires secure contexts but works fine for me on localhost http.

Maybe Bluetooth has particularly stringent requirements 🤷 

Anyway, I'll play around with this when I get to testing. The main considerations are:
- Ensuring this is an actual problem, and
- Ensuring the solution works easily and cross-platform with minimal dependencies

**Review by albertnis — COMMENTED** [review]
Thanks for the PR! I appreciate your contribution. I'm certainly interested in this feature (and frankly didn't know that these sensors could be configured over bluetooth). Might be a week or two before I can spare time to test and do a thorough review of your implementation.

Apart from the connection method itself, are there any differences between configuring over uart vs BT? For example: Any configuration features which are disabled when using BT? Your implementation looks like one can send and receive all the same payloads, which would be ideal.

**Comment by belens** [comment]
> Thanks for the PR! I appreciate your contribution. I'm certainly interested in this feature (and frankly didn't know that these sensors could be configured over bluetooth). Might be a week or two before I can spare time to test and do a thorough review of your implementation.
> 
> Apart from the connection method itself, are there any differences between configuring over uart vs BT? For example: Any configuration features which are disabled when using BT? Your implementation looks like one can send and receive all the same payloads, which would be ideal.

based on the hardware docs (LD2410C) the only difference is that you have to send a login message, and it will not respond in the same way if you try the same over serial. see "2.2.14 Obtaining bluetooth permissions"
https://docs.google.com/document/d/1qd5nMgZ7EcKPAqaGAXYy7aXqgPdciiP5WibcO418FpU/edit

Do we maybe have different versions of the same hardware?

File: vite.config.ts (hunk 2 above)
Comment (belens) [reply vite.config.ts:6]: You are probably right. I'll look into this further for you maybe it is not needed anymore. I've had to do a strange setup on my end. I didn't have a working bluetooth implementation yet, and MacOS serial connection was giving me a broken datastream so I ended up attaching the hardware to a windows pc (so I can test with the working windows serial data stream) that visited the website over local ip 192.168.0.185:80 (which i was running from my macbook) and that forced me to implement https if that makes sense.

tldr; I can probably remove it.

File: .vscode/settings.json (hunk 1 above)
Comment (belens) [reply .vscode/settings.json:19]: I'll refactor this, thanks for the feedback!

File: src/lib/components/connectionScreen.svelte [hunk 3]
```diff
@@ -57,5 +62,11 @@
       class="hover:bg-blue-500 active:bg-blue-700 bg-blue-600 border rounded-sm border-blue-500 px-2 py-1"
       on:click={onClick}>Select port and connect...</button
     >
+
+
+    <button
+      class="hover:bg-blue-500 active:bg-blue-700 bg-blue-600 border rounded-sm border-blue-500 px-2 py-1"
+      on:click={openBluetooth}>BLUETOOTH</button
```
Comment (albertnis) [inline src/lib/components/connectionScreen.svelte:69]: Let's align this button with the other one:

- The existing button text could be renamed to "Connect over serial…" and this one to "Connect over Bluetooth…"
- Rename existing `onClick` to `onClickSerial` - then call `onClickBluetooth` from here (it's currently unused). This could be an opportunity to improve the exported callbacks themselves: `connectSerial` and `connectBluetooth` instead of `openPort` and `openBluetooth`.

File: vite.config.ts (hunk 2 above)
Comment (albertnis) [reply vite.config.ts:6]: I've tested this and it works fine with localhost as a secure context. Therefore, you should be good to revert this file and remove the `mkcert` dependency.

File: src/lib/stores/serialRead.ts [hunk 4]
```diff
@@ -99,7 +99,7 @@ export const createSerialReadStore = (
               everyEqual(inputBytes.slice(-4), radarDataOutputPayloadTrailer) ||
               everyEqual(inputBytes.slice(-4), configurationPayloadTrailer)
             ) {
-              const payload = new Uint8Array(inputBytes);
+              const payload = new Uint8Array(inputBytes);
```
Comment (albertnis) [inline src/lib/stores/serialRead.ts:102]:
```suggestion
              const payload = new Uint8Array(inputBytes);
```

File: package.json [hunk 5]
```diff
@@ -32,7 +32,9 @@
     "tailwindcss": "^3.3.3",
     "tslib": "^2.6.2",
     "typescript": "^5.2.2",
-    "vite": "^4.4.11"
+    "vite": "^4.4.11",
+    "vite-plugin-mkcert": "^1.17.5",
+    "webbluetooth": "^3.2.1"
```
Comment (albertnis) [inline package.json:37]: `webbluetooth` is a node implementation but we should be able to just use the web APIs here (as you already are).

So we should remove it.

File: src/lib/components/app.svelte [hunk 6]
```diff
@@ -3,25 +3,58 @@
...
+  const openBluetooth = async () => {
+    console.log("openBluetooth");
+
+    try {
+      const device = await (
+        navigator as Navigator & { bluetooth: any }
+      ).bluetooth.requestDevice({
+        filters: [
+          {
+            name: "HLK-LD2410_6F1F", // TODO: make this configurable
```
Comment (albertnis) [inline src/lib/components/app.svelte:29]: `namePrefix` should work here. The following works for me on my LD2410C. Would be nice to check it works for a LD2410B but I haven't got mine wired up yet.

```suggestion
            namePrefix: "HLK-LD2410",
```

File: src/lib/components/connectionScreen.svelte [hunk 7]
```diff
@@ -57,5 +62,11 @@
       class="hover:bg-blue-500 active:bg-blue-700 bg-blue-600 border rounded-sm border-blue-500 px-2 py-1"
       on:click={onClick}>Select port and connect...</button
     >
+
+
+    <button
```
Comment (albertnis) [inline src/lib/components/connectionScreen.svelte:67]: We should disable this button if `window.navigator === undefined`.
There's probably some broader improvements around connection instructions and compatibility checking now that there's a second connection mechanism, but I'm happy to tackle that separately later.

File: src/lib/stores/bluetoothRead.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+interface Store {
+  subscribe: (subscription: SubscribeCallback) => () => void;
+}
+
+export type SerialStore = Store & {
+  connect: () => void;
+  disconnect: () => void;
+  write: (payload: Uint8Array) => void;
+};
```
Comment (albertnis) [inline src/lib/stores/bluetoothRead.ts:50]: At a glance, these types are all duplicate with `serialRead.ts`. I'd suggest sharing them. Maybe the following directory structure could work.

```
src
   lib
      stores
         serial
            bluetooth.ts
            serial.ts   <- renamed and moved from serialRead.ts
            types.ts    <- shared types in here
```

`serialWrite.ts` can also be deleted.

File: src/lib/components/app.svelte [hunk 9]
```diff
@@ -3,25 +3,58 @@
...
-    });
+    serialReadStore.subscribe(handleNotification);
+  };
+
+  const openBluetooth = async () => {
+    console.log("openBluetooth");
+
+    try {
+      const device = await (
+        navigator as Navigator & { bluetooth: any }
```
Comment (albertnis) [inline src/lib/components/app.svelte:25]: I'd prefer to use [`@types/web-bluetooth`](https://www.npmjs.com/package/@types/web-bluetooth) instead of `any`s. Hopefully TypeScript gets built-in support for web Bluetooth but doesn't look like it's happened yet.

File: src/lib/stores/bluetoothRead.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+export const createBluetoothReadStore = (
+  server: any
+): SerialStore => {
+  let stopping = false;
+  let sendPassCharacteristic: any | null = null;
+  let readDataCharacteristic: any | null = null;
```
Comment (albertnis) [inline src/lib/stores/bluetoothRead.ts:57]: With the types library I suggested above:

```suggestion
  server: BluetoothRemoteGATTServer
): SerialStore => {
  let stopping = false;
  let sendPassCharacteristic: BluetoothRemoteGATTCharacteristic | null = null;
  let readDataCharacteristic: BluetoothRemoteGATTCharacteristic | null = null;
```

and go from there...

File: src/lib/stores/bluetoothRead.ts [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        broadcastEvent({
+          eventType: "READ",
+          payload: new Uint8Array(event.target.value.buffer),
+        });
+      }
+    );
+  };
+  const login = async () => {
+    // this is hex version of the login command with the default password
+    const hexString = "FDFCFBFA0800A80048694C696E6B04030201";
```
Comment (albertnis) [inline src/lib/stores/bluetoothRead.ts:95]: This is probably fine as I imagine that most people are just leaving the default password. It would be nice to add a way to connect with a custom password someday (and maybe even change it), but I'm happy to leave that out of scope.

File: src/lib/stores/bluetoothRead.ts [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    readForever();
+    writeForever();
+    broadcastEvent({
+      eventType: "CONNECT",
+    });
+  };
+
+  const disconnect = async () => {
+    stopping = true;
+    server.disconnect(); // TODO: refactor
```
Comment (albertnis) [inline src/lib/stores/bluetoothRead.ts:116]: We need to emit a disconnection event. I think it should be fine to simply inline it here

```suggestion
    server.disconnect(); // TODO: refactor
    broadcastEvent({
      eventType: "DISCONNECT",
    });
```

It's a fair bit cleaner than serial disconnection. 

Can probably remove the `stopping` concept from this file, too.

File: src/lib/components/app.svelte [hunk 13]
```diff
@@ -3,25 +3,58 @@
...
+      bluetoothReadStore.subscribe(handleNotification);
+    } catch (error) {
+      console.error("Argh! " + error);
+    }
+  };
+
+  const handleNotification = (event) => {
+    if (event.eventType === "DISCONNECT") {
+      // serialReadStore = null; // to avoid the page dissapearing
+    }
```
Comment (albertnis) [inline src/lib/components/app.svelte:48]: What works for me is to clear both stores then move the definition above both `openPort` and `openBluetooth`

```suggestion
  const handleNotification = (event) => {
    if (event.eventType === "DISCONNECT") {
      serialReadStore = null;
      bluetoothReadStore = null;
    }
```

File: src/lib/stores/bluetoothRead.ts [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+  };
+
+  const readForever = async () => {
+    if (!readDataCharacteristic) { return }
+    await readDataCharacteristic.startNotifications();
+    readDataCharacteristic.addEventListener(
+      "characteristicvaluechanged",
+      (event: any) => {
+        broadcastEvent({
+          eventType: "READ",
```
Comment (albertnis) [inline src/lib/stores/bluetoothRead.ts:87]: I'm seeing an odd bug where if I connect to bluetooth, disconnect, then reconnect, there's no output in the console component. It's always the second connection on bluetooth; serial doesn't have the same issue.

Looks like `outputDiv` in  `bind:this={outputDiv}` gets weirdly set to `null`. It's likely an existing bug and maybe even a bug from Svelte itself, so probably fine to leave a fix out of scope unless you have any ideas.

**Review by albertnis — COMMENTED** [review]
Months later, I got around to testing and reviewing this. The functionality is really nice - thanks so much for your work on this. Most of my comments are just around factoring and the odd little thing.

**Comment by albertnis** [comment]
Superseded by #16
